### ISSUE TITLE
[Bug]: Plenty of RecaptchaException Internal Error and Network Error

### ISSUE THREAD

**Issue by cj3g10** [GoogleCloudPlatform/recaptcha-enterprise-mobile-sdk#80]
### Are you seeing the problem on Android or iOS?

Android and iOS

### Describe the bug

getClient and execute returns Internal Errors or Network Errors

### Site Key

Can provide if requested

### Select the method used to integrate with reCAPTCHA Mobile.

Cocoapods (iOS)
GMaven (Android)

### SDK Version

18.4.0

### To Reproduce

We cannot reproduce the issue, as during our development and testing, the recaptcha is working properly and it still does for us. But we are getting reports that a lot of our users cannot log in because of the recaptcha. After logging the errors to our crashlytics, we see that there are over 158k events of RecaptchaException logged over the past 7 days.

### Development IDE

Android Studio Hedgehog, XCode 15.1

### To Reproduce

There doesn't seem to be a pattern for what devices it occurs on.
This happens on iOS and Android, with a majority of the Android devices being Samsung at 91%.

### Relevant log output

```shell
Non-fatal Exception: com.google.android.recaptcha.RecaptchaException
Network Error
com.google.android.recaptcha.internal.zzbq.zza (com.google.android.recaptcha:recaptcha@@18.4.0:22)
com.google.android.recaptcha.internal.zzal.invokeSuspend (com.google.android.recaptcha:recaptcha@@18.4.0:13)
kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (ContinuationImpl.kt:33)
kotlinx.coroutines.DispatchedTask.run (DispatchedTask.kt:108)
kotlinx.coroutines.internal.LimitedDispatcher$Worker.run (LimitedDispatcher.kt:115)
kotlinx.coroutines.scheduling.TaskImpl.run (Tasks.kt:103)
kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely (CoroutineScheduler.java:584)
kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask (CoroutineScheduler.kt:793)
kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker (CoroutineScheduler.kt:697)
kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run (CoroutineScheduler.kt:684) 

Non-fatal Exception: com.google.android.recaptcha.RecaptchaException
Internal Error
com.google.android.recaptcha.internal.zzam.zza (com.google.android.recaptcha:recaptcha@@18.4.0:27)
com.google.android.recaptcha.internal.zzai.invokeSuspend (com.google.android.recaptcha:recaptcha@@18.4.0:20)
kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (ContinuationImpl.kt:33)
kotlinx.coroutines.internal.ScopeCoroutine.afterResume (Scopes.kt:32)
kotlinx.coroutines.AbstractCoroutine.resumeWith (AbstractCoroutine.kt:102)
kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (ContinuationImpl.kt:46)
kotlinx.coroutines.internal.ScopeCoroutine.afterResume (Scopes.kt:32)
kotlinx.coroutines.AbstractCoroutine.resumeWith (AbstractCoroutine.kt:102)
kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (ContinuationImpl.kt:46)
kotlinx.coroutines.DispatchedTask.run (DispatchedTask.kt:106)
android.os.Handler.handleCallback (Handler.java:942)
android.os.Handler.dispatchMessage (Handler.java:99)
android.os.Looper.loopOnce (Looper.java:226)
android.os.Looper.loop (Looper.java:313)
android.app.ActivityThread.main (ActivityThread.java:8762)
java.lang.reflect.Method.invoke (Method.java)
com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:604)
com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1067)
```

### Relevant log output

![Screenshot 2023-12-28 at 6 09 48 PM](https://github.com/GoogleCloudPlatform/recaptcha-enterprise-mobile-sdk/assets/5802544/a34c8921-0d18-4126-9b54-f889f6d1d33b)

**Comment by mcorner** [comment]
- We are working to drive "internal error" to 0 in Q1, but network errors are exactly that (it can't contact our server).
- What is your sitekey(s)?  Then I can look at the data in the backend.
- What rate/percentage of each kind of exception are you seeing?  
- Have you implemented retries? (see the example code in the repo)
- We never suggest outright blocking users based on reCAPTCHA and instead recommend using other sources of friction (MFA etc).

**Comment by cj3g10** [comment]
Android: 6LcwdxgnAAAAAP2muhaFiSjUJfFkLtqfrg7P0WTi
iOS: 6LfYrBgnAAAAAKoMDDXztqyJr4B2dc-7qKiZwu2k

It's kind of hard for us to give a percentage of each kind of exception as Crashlytics is grouping all the Issues by it's stack trace even when they have different exceptions. All I can say is majority of them are internal errors or network errors, with a few sprinkles of Invalid KeyType, which I have no idea how this exception is thrown since the Key Type is most certainly correctly.

We have implemented trying to getClient again before user attempts an execute if the previous attempt to getClient has failed.

My primary goal is to understand why we're getting so many exceptions and why it happens to some users and not to the others. Obviously some of these exceptions may be the fault of our end users, but it's kind of hard to tell our client that we don't know why the recaptcha is throwing these exceptions while Crashlytics is logging over 156k events and 48k unique users.

**Comment by mcorner** [comment]
The internal error and key type error (off play devices trying to use your app when you only have an on play key) errors are a tiny fraction of your inits and errors.  Almost 100% of the errors are network related.  They may be somewhat better on 18.4.0 than your previous version (we reworked the networking in 18.4.0).  

But let's see if we can track down the source of the network errors.   

Do you check for a network connection before calling recaptcha?  Perhaps try that, log it, and see what you get: https://github.com/GoogleCloudPlatform/recaptcha-enterprise-mobile-sdk/blob/main/examples/android/basic/app/src/main/java/com/google/recaptcha/sample/NetworkUtils.kt#L102

The SDK has to be able to reach recaptcha.net. Is there some possibility that host is blocked for your users?

You can log this as well: InetAddress.getByName("www.recaptcha.net").isReachable(10_000)

**Comment by cj3g10** [comment]
Thanks for exception rate info. This will be useful for our further investigation.

I don't think it's a network connection issue for most users. Our client's analytics tools can see that other network requests are working fine when only recaptcha is 400.
![image](https://github.com/GoogleCloudPlatform/recaptcha-enterprise-mobile-sdk/assets/5802544/ec34b02d-85ca-48aa-ba10-20ca95fd19bd)

We can try logging for the reachability, although we won't have any results to show until we submit a new release.

What reasons could there be for the recaptcha host being blocked by our users? There isn't a pattern I can see in our analysis data that can see that would show what would cause the recaptcha host to be blocked.

**Comment by mcorner** [comment]
Do you have any data around the failed requests to /mri ? (that is reCAPTCHA initializing).  Is it a 400 for sure?  You will get 400s for wrong key type (off play devices), but it shouldn't happen very often (looks like a few times per day in your case).  If it is happening when you are getting network error, we need to dig into this.

Also, contact me at my firstname plus my lastname at google.com to make sure we figure this out.  We will dig into it, but it will be a bit slow due to the holidays here.  Please leave reCAPTCHA running, but you shouldn't be blocking on missing tokens for now.

Thanks for all of the good data.

-Mark

**Comment by mcorner** [comment]
@rogu01 We will need more information to help you effectively. Sitekey and what errors and and what rates would be helpful to start.

**Comment by rogu01** [comment]
Hi @cj3g10 
Did you manage to solve this issue? I was wondering if you had any suggestions on how to fix it.

**Comment by mcorner** [comment]
@romaCaldeira please open another issue and fill out the issue form, this sounds like something else

**Comment by romaCaldeira** [comment]
> @romaCaldeira please open another issue and fill out the issue form, this sounds like something else

I believe it is the same problem, do you have any predictions for fixing the problem mentioned by @cj3g10

**Comment by mcorner** [comment]
@romaCaldeira we can determine if it is the same issue or not after you open a new issue. When you open a new issue it will ask you for important details like your sitekey that really helps us dig into the data.  Thanks.

**Comment by mcorner** [comment]
The original issue contained two parts: network errors and internal errors. We believe we have addressed the internal errors in 18.5.0-beta02. Please give that a try. 

Networking errors have been traced to calling the API when there is no network (and confirmed by Android apis). The sdk must have networking while initializing and the only mitigation at the moment is to retry when it does. We are working on a new API to internalize these retries (likely in 18.6). 

So there isn't a "bug" but we recognize it could be easier to use and we are working hard.on that (next few months).
